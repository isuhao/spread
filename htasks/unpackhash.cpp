#include "unpackhash.hpp"
#include "tasks/unpack.hpp"
#include <mangle/vfs/stream_factory.hpp>
#include <mangle/stream/servers/null_stream.hpp>
#include "hash/hash_stream.hpp"
#include <boost/filesystem.hpp>
#include <stdexcept>

using namespace Spread;
using namespace Mangle::Stream;

//#define DEBUG_PRINT
#ifdef DEBUG_PRINT
#include <iostream>
#define PRINT(a) std::cout << __LINE__ << ": " << a << "\n";
#else
#define PRINT(a)
#endif

/* This output writer is used for indexing. It stores names + hashes
   in an index, and can optionally write data to files.
 */

struct UH_ListHasher : Mangle::VFS::StreamFactory
{
  Hash::DirMap *index;

  HashStreamPtr stream;
  std::string lastName;
  bool lastDir;
  std::string where;
  bool absPaths;

  ~UH_ListHasher() { open(""); }

  // This follows the same close-then-open mechanics as
  // getOutStream()/closeStream() in HashTask.
  StreamPtr open(const std::string &name)
  {
    PRINT("open(" << name << ")  lastName=" << lastName);

    // Store the last hash, if any
    if(stream && lastName != "")
      {
        // Ignore directories
        if(!lastDir)
          (*index)[lastName] = stream->finish();
        lastName = "";
      }

    // Allow an empty name to close the last file (above) without
    // opening a new one here.
    if(name == "") return StreamPtr();

    // Is this a directory?
    char ch = name[name.size()-1];
    lastDir = (ch == '/' || ch == '\\');
    lastName = name;

    if(where == "")
      {
        // If we're not writing anything, just reuse a NullStream as
        // the final output.
        if(!stream)
          stream.reset(new HashStream(StreamPtr(new NullStream)));
      }
    else
      {
        using namespace boost::filesystem;
        path abs = where;
        abs /= name;

        PRINT("  abs=" << abs);

        if(absPaths)
          lastName = abs.string();

        stream.reset();
        if(!lastDir)
          {
            create_directories(abs.parent_path());
            stream.reset(new HashStream(abs.string(), true));
          }
        else
          {
            create_directories(abs);
          }
      }

    if(lastDir) return StreamPtr();
    return stream;
  }
};

void UnpackHash::makeIndex(const std::string &arcFile, Hash::DirMap &index,
                           const std::string &where)
{
  UH_ListHasher *m = new UH_ListHasher;
  m->index = &index;
  m->where = where;
  m->absPaths = false;
  Mangle::VFS::StreamFactoryPtr mp(m);

  UnpackTask unp(arcFile, mp);
  unp.run();
  // Needed to include the last file
  m->open("");

  unp.failError();
}

/* This output writer is used for unpacking. It looks up filenames and
   finds their hashes.
 */
struct UH_ListUser : Mangle::VFS::StreamFactory
{
  Hash::DirMap index;
  HashTask *owner;

  UH_ListUser(HashTask *t) : owner(t) { assert(owner); }

  StreamPtr open(const std::string &name)
  {
    // Look up the hash from the name
    Hash h = index[name];

    // Fail if the unpacker threw an unknown file at us
    if(h.isNull())
      throw std::runtime_error("Unexpected file in archive: " + name);

    // HashTask takes care of the rest
    return owner->getOutStream(h);
  }
};

Job *UnpackHash::createJob()
{
  // Get the input filename
  assert(inputs.size() == 1);
  std::string file = inputs.begin()->second;

  PRINT("UnpackHash::createJob: file=" << file);

  // Figure out if we are doing a "blind" unpack
  if(blindDir != "")
    {
      PRINT("  Blind unpack to: " << blindDir << " absPahts=" <<
            (absPaths?"true":"false"));

      UH_ListHasher *m = new UH_ListHasher;
      assert(blindOut);
      m->index = blindOut;
      m->where = blindDir;
      m->absPaths = absPaths;
      Mangle::VFS::StreamFactoryPtr mp(m);

      PRINT("Returning job");
      return new UnpackTask(file, mp);
    }

  // The rest is for non-blind unpacking

  UH_ListUser *m = new UH_ListUser(this);
  Mangle::VFS::StreamFactoryPtr mp(m);

  // Copy requested items from the full index over to the working
  // index
  Hash::DirMap::iterator it;

  // Just to optimize a tiny bit more - use this to prune repeated
  // hashes from the file list
  std::set<Hash> hashes;

  // Loop through the zip index and only pick the files we actually
  // want to use for unpacking.
  for(it = index.begin(); it != index.end(); ++it)
    {
      const std::string &name = it->first;
      const Hash &hsh = it->second;

      // Do we need this object?
      if(outputs.find(hsh) == outputs.end())
        // Hash was not found in the requested output list, skip it.
        continue;

      // Have we already listed this hash?
      if(hashes.count(hsh) != 0)
        // Prune duplicates
        continue;

      // Is the hash empty? Might be the case for directories. If so,
      // skip it.
      if(hsh.isNull())
        continue;

      /* This file is a keeper. Add it to the file list, as well as to
         the live lookup index.
       */
      list.insert(name);
      m->index[name] = hsh;
      hashes.insert(hsh);
    }

  // Set up the unpacking job
  return new UnpackTask(file, mp, &list);
}
