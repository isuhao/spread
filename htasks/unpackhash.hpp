#ifndef __HASH_UNPACKTASK_HPP_
#define __HASH_UNPACKTASK_HPP_

#include "hashtask.hpp"
#include <set>

/* This class unpacks one archive file, and writes outputs.

   - outputs is set as normal through HashTask::addOutput()
   - inputs must list exactly one archive file
   - index must be pre-generated with makeIndex()
 */

namespace Spread
{
  struct UnpackHash : HashTask
  {
    typedef std::map<std::string, Hash> HashMap;
    typedef std::set<std::string> FileList;

    /* The parameter is the index used to look up the archive file. The
       string format is archive-type specific, so you should only
       supply an index created with makeIndex().

       Don't worry if the index contains more files than you need to
       extract. Files not requested with HashTask::addOutput() will
       still be ignored.
     */
    HashMap index;
    UnpackHash() {}
    UnpackHash(const HashMap &_index) : index(_index) {}

    /* Generate an index from an archive file.

       An optional dir 'where' can be used to specify an output
       directory. If empty, no files are written.
     */
    static void makeIndex(const std::string &arcFile, HashMap &index,
                          const std::string &where = "");

  private:
    Job *createJob();
    FileList list;
  };
};

#endif
