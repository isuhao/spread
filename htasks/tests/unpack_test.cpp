#include "unpackhash.hpp"
#include <iostream>
#include "print_dir.hpp"

using namespace std;
using namespace Spread;

typedef Hash::DirMap HMap;

void print(const HMap &mp)
{
  cout << "Index:\n";
  HMap::const_iterator it;
  for(it = mp.begin(); it != mp.end(); it++)
    {
      cout << "  " << it->second << "   - " << it->first << endl;
    }
}

string D1 = "hey\n";
string D2 = "yo man\n";
Hash H1(D1.c_str(), D1.size());
Hash H2(D2.c_str(), D2.size());

int main()
{
  UnpackHash unp;

  UnpackHash::makeIndex("test1.zip", unp.index, "_unp_test2");
  print(unp.index);

  unp.addInput(Hash(), "test1.zip");
  unp.addOutput(H2, "_unp_test1/file2.txt");
  unp.addOutput(H2, "_unp_test1/file2_copy.txt");
  unp.addOutput(H1, "_unp_test1/dir/to/file1.txt");
  unp.addOutput(H2, "_unp_test1/file2_again.txt");

  unp.run();

  printDir("_unp_test1");
  printDir("_unp_test2");

  return 0;
}
