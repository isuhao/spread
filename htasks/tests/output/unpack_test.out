5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       a
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       b
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       c
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       d
TpVf6gJoUYy6pQBAnfvsiPDs660o2E7L4lC67ZfbqIkE       dir/file1.txt
nNdHCq2-eBiq_4tNjqrg-9lTQtdcW4fynNawsk8q_FAW       dir/file3.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       e
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       f
TpVf6gJoUYy6pQBAnfvsiPDs660o2E7L4lC67ZfbqIkE       file1_again.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       file2.txt
VFJt-Q1dairWD3TFtxh08OpLdGk4XKaqd4SRhfH8KC8V       file4.txt
Total 11 elements
Hash: qMwIi1GuxuK3mFPB41SNXnOpXBUGGz0kEipHdvLyiGwXAg

Directory: _unp_test1
00                                                 dir/
00                                                 dir/to/
TpVf6gJoUYy6pQBAnfvsiPDs660o2E7L4lC67ZfbqIkE       dir/to/file1.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       file2.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       file2_again.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       file2_copy.txt
Total 6 elements
Hash: lwvjrlGl_InhIIsZV-slIxqrlpazlWLKCzej5AuIOyxFAQ

Directory: _unp_test2
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       a
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       b
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       c
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       d
00                                                 dir/
TpVf6gJoUYy6pQBAnfvsiPDs660o2E7L4lC67ZfbqIkE       dir/file1.txt
nNdHCq2-eBiq_4tNjqrg-9lTQtdcW4fynNawsk8q_FAW       dir/file3.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       e
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       f
TpVf6gJoUYy6pQBAnfvsiPDs660o2E7L4lC67ZfbqIkE       file1_again.txt
5Kgp14e_O89BLn-fGjbNzchGsPkARpFXrlGSAWZScVQH       file2.txt
VFJt-Q1dairWD3TFtxh08OpLdGk4XKaqd4SRhfH8KC8V       file4.txt
Total 12 elements
Hash: oFevl728ZpfLmg8jGI7XvughikkqvDmyZ6eLJ4T_muNFAg
